debian/mongotop.1
.\" Man page generated from reStructuredText.
.
.TH "MONGOTOP" "1" "Oct 24, 2019" "4.0" "mongodb-manual"
.SH NAME
mongotop \- MongoDB Activity Monitor
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Description\fP
.IP \(bu 2
\fI\%Syntax\fP
.IP \(bu 2
\fI\%Required Access\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Fields\fP
.IP \(bu 2
\fI\%Examples\fP
.IP \(bu 2
\fI\%Additional Information\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.IP "macOS Sierra and Go 1.6 Incompatibility"
.sp
Users running on macOS Sierra require the 3.2.10 or newer version
of  mongotop\&.
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
\fI\%mongotop\fP provides a method to track the amount of time a
MongoDB instance \fBmongod\fP spends reading and writing data.
\fI\%mongotop\fP provides statistics on a per\-collection level.
By default, \fI\%mongotop\fP returns values every second.
.sp
Run \fI\%mongotop\fP from the system command line, not the \fBmongo\fP shell.
.SH SYNTAX
.SS Default
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop [frequency]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can run \fI\%mongotop\fP without any command\-line options to
connect to the \fBmongod\fP instance running on the localhost
port \fB27017\fP\&. By default, \fI\%mongotop\fP returns values
every second.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
To have \fI\%mongotop\fP report every 30 seconds, specify the
time:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop 30
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
When connected to a \fBmongod\fP instance, the program reports
first on the connection and then reports on the statistics at the
configured frequency.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
2019\-04\-29T15:35:27.785\-0400 connected to: 127.0.0.1

             ns    total      read      write    <timestamp>
<db.collection>    81802ms     0ms    81802ms
\&...
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Command Line Options
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop [options] [frequency]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can run \fI\%mongotop\fP with various \fI\%command\-line
options\fP\&.
.sp
For example, to connect \fI\%mongotop\fP to a
\fBmongod\fP instance running on a remote host
\fBmongodb0.example.com\fP and report every 30 seconds:
.INDENT 0.0
.IP \(bu 2
You can include the desired frequency of reporting (in this
example, 30 seconds) and the \fI\%\-\-uri\fP
option to specify the host and port:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop 30 \-\-uri mongodb://mongodb0.example.com:27017 [additional options]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If the \fBmongod\fP instance requires authentication, you
can specify the user, password, and authentication database as part
of the URI connection string:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop 30 \-\-uri mongodb://user:password@mongodb0.example.com:27017/?authSource=admin [additional options]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The user running \fI\%mongotop\fP must have the
\fBserverStatus\fP and \fBtop\fP privileges.
.sp
For more information on the URI connection string, see
\fI\%\-\-uri\fP\&.
.IP \(bu 2
Alternatively, you can use the \fI\%\-\-host\fP
and \fI\%\-\-port\fP options to specify the host
and port:
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop 30 \-\-host mongodb0.example.com \-\-port 27017 [additional options]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If the \fBmongod\fP instance requires authentication, you
can specify the user \fI\%\-u\fP, password
\fI\%\-p\fP, and the authentication database
\fI\%\-\-authenticationDatabase\fP\&.
.INDENT 2.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop 30 \-\-host mongodb0.example.com \-\-port 27017 \-u user \-p password \-\-authenticationDatabase admin [additional options]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The user running \fI\%mongotop\fP must have the
\fBserverStatus\fP and \fBtop\fP privileges.
.UNINDENT
.sp
For more information on the options available, see
\fI\%Options\fP\&.
.sp
When connected to a \fBmongod\fP instance, the program reports
first on the connection and then reports on the statistics at the
configured frequency.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
2019\-04\-29T15:35:27.785\-0400 connected to: 127.0.0.1

             ns    total      read      write    <timestamp>
<db.collection>    81802ms     0ms    81802ms
\&...
.ft P
.fi
.UNINDENT
.UNINDENT
.SH REQUIRED ACCESS
.sp
In order to connect to a \fBmongod\fP that enforces
authorization, the connecting user must have the
\fBserverStatus\fP and \fBtop\fP privileges.
.sp
The built\-in role \fBclusterMonitor\fP provides these privileges.
.SH OPTIONS
.INDENT 0.0
.TP
.B mongotop
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBmongotop\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs \fBmongotop\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
output from database commands
.IP \(bu 2
replication activity
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmongotop\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-uri <connectionString>
New in version 3.4.6.

.sp
Specify a resolvable URI connection string to connect to the MongoDB deployment.
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-\-uri "mongodb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]"
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For more information on the components of the connection string, see
the Connection String URI Format documentation.
.sp
\fBIMPORTANT:\fP
.INDENT 7.0
.INDENT 3.5
The following command\-line options cannot be used in conjunction
with \fI\%\-\-uri\fP option:
.INDENT 0.0
.IP \(bu 2
\fI\%\-\-host\fP
.IP \(bu 2
\fI\%\-\-port\fP
.IP \(bu 2
\fI\%\-\-username\fP
.IP \(bu 2
\fI\%\-\-password\fP (if the
URI connection string also includes the password)
.IP \(bu 2
\fI\%\-\-authenticationDatabase\fP
.IP \(bu 2
\fI\%\-\-authenticationMechanism\fP
.UNINDENT
.sp
Instead, specify these options as part of your \fI\%\-\-uri\fP
connection string.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>, \-h <hostname><:port>
\fIDefault\fP: localhost:27017
.sp
Specifies a resolvable hostname for the \fBmongod\fP to which to
connect. By default, the \fBmongotop\fP attempts to connect to a MongoDB
instance running on the localhost on port number \fB27017\fP\&.
.sp
To connect to a replica set, specify the
\fBreplSetName\fP and a seed list of set members, as in
the following:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-\-host <replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
When specifying the replica set list format, \fBmongotop\fP always connects to
the primary\&.
.sp
You can also connect to any single member of the replica set by specifying
the host and port of only that member:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
\-\-host <hostname1><:port>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Changed in version 3.0.0: If you use IPv6 and use the \fB<address>:<port>\fP format, you must
enclose the portion of an address and port combination in
brackets (e.g. \fB[<address>]\fP).

.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
You cannot specify both \fI\%\-\-host\fP and \fI\%\-\-uri\fP\&.
.UNINDENT
.UNINDENT
.sp
If connected to a replica set where the primary is not
reachable, \fBmongotop\fP returns an error message.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
\fIDefault\fP: 27017
.sp
Specifies the TCP port on which the MongoDB instance listens for
client connections.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
You cannot specify both \fI\%\-\-port\fP and \fI\%\-\-uri\fP\&.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
\fIRemoved in version 3.0.\fP
.sp
Enables IPv6 support and allows \fBmongotop\fP to connect to the
MongoDB instance using an IPv6 network. Prior to MongoDB 3.0, you
had to specify \fI\%\-\-ipv6\fP to use IPv6. In MongoDB 3.0 and later, IPv6
is always enabled.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBmongod\fP or \fBmongos\fP that has
TLS/SSL support enabled.
.sp
For more information about TLS/SSL and MongoDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
Starting in version 3.4, if \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP is not
specified and you are not using x.509 authentication, the
system\-wide CA certificate store will be used when connecting to an
TLS/SSL\-enabled server.
.sp
If using x.509 authentication, \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP
must be specified unless using \fB\-\-sslCertificateSelector\fP\&.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
\fBVersion 3.2 and earlier:\fP For TLS/SSL connections (\fB\-\-ssl\fP) to
\fBmongod\fP and \fBmongos\fP, if the \fBmongotop\fP runs without the
\fI\%\-\-sslCAFile\fP, \fBmongotop\fP will not attempt
to validate the server certificates. This creates a vulnerability
to expired \fBmongod\fP and \fBmongos\fP certificates as
well as to foreign processes posing as valid \fBmongod\fP or
\fBmongos\fP instances. Ensure that you \fIalways\fP specify the
CA file to validate the server certificates in cases where
intrusion is a possibility.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongoDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the TLS/SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\%\-\-ssl\fP option to connect
to a \fBmongod\fP or \fBmongos\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
For more information about TLS/SSL and MongoDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fI\%\-\-sslPEMKeyFile\fP). Use the \fI\%\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBmongotop\fP will
redact the password from all logging and reporting output.
.sp
If the private key in the PEM file is encrypted and you do not specify
the \fI\%\-\-sslPEMKeyPassword\fP option, the \fBmongotop\fP will prompt for a passphrase. See
ssl\-certificate\-password\&.
.sp
For more information about TLS/SSL and MongoDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
For more information about TLS/SSL and MongoDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, MongoDB logs as a
warning the use of the invalid certificate.
.sp
Starting in MongoDB 4.0, if you specify
\fB\-\-sslAllowInvalidCertificates\fP or \fBssl.allowInvalidCertificates:
true\fP when using x.509 authentication, an invalid certificate is
only sufficient to establish a TLS/SSL connection but is
\fIinsufficient\fP for authentication.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
Although available, avoid using the
\fB\-\-sslAllowInvalidCertificates\fP option if possible. If the use of
\fB\-\-sslAllowInvalidCertificates\fP is necessary, only use the option
on systems where intrusion is not possible.
.sp
If the \fBmongo\fP shell (and other
mongodb\-tools\-support\-ssl) runs with the
\fB\-\-sslAllowInvalidCertificates\fP option, the
\fBmongo\fP shell (and other
mongodb\-tools\-support\-ssl) will not attempt to validate
the server certificates. This creates a vulnerability to expired
\fBmongod\fP and \fBmongos\fP certificates as
well as to foreign processes posing as valid
\fBmongod\fP or \fBmongos\fP instances. If you
only need to disable the validation of the hostname in the
TLS/SSL certificates, see \fB\-\-sslAllowInvalidHostnames\fP\&.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongoDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in TLS/SSL certificates. Allows
\fBmongotop\fP to connect to MongoDB instances even if the hostname in their
certificates do not match the specified hostname.
.sp
For more information about TLS/SSL and MongoDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBmongotop\fP to use the FIPS mode of the TLS/SSL
library. Your system must have a FIPS compliant library to use
the \fI\%\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS\-compatible TLS/SSL is
available only in \fI\%MongoDB Enterprise\fP\&. See
/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a MongoDB database
that uses authentication. Use in conjunction with the \fI\%\-\-password\fP and
\fI\%\-\-authenticationDatabase\fP options.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
You cannot specify both \fI\%\-\-username\fP and \fI\%\-\-uri\fP\&.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a MongoDB database
that uses authentication. Use in conjunction with the \fI\%\-\-username\fP and
\fI\%\-\-authenticationDatabase\fP options.
.sp
Changed in version 3.0.2: To prompt the user
for the password, pass the \fI\%\-\-username\fP option without
\fI\%\-\-password\fP or specify an empty string as the \fI\%\-\-password\fP value,
as in \fB\-\-password ""\fP .

.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
You cannot specify both \fI\%\-\-password\fP and \fI\%\-\-uri\fP\&.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationDatabase <dbname>
Specifies the authentication database where the specified \fI\%\-\-username\fP has been created.
See user\-authentication\-database\&.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
You cannot specify both \fI\%\-\-authenticationDatabase\fP and \fI\%\-\-uri\fP\&.
.UNINDENT
.UNINDENT
.sp
Changed in version 3.0.0: \fI\%\-\-authenticationDatabase\fP is required for \fBmongod\fP
and \fBmongos\fP instances that use authentication\&.

.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationMechanism <name>
\fIDefault\fP: SCRAM\-SHA\-1
.sp
Specifies the authentication mechanism the \fBmongotop\fP instance uses to
authenticate to the \fBmongod\fP or \fBmongos\fP\&.
.sp
Changed in version 4.0: MongoDB removes support for the deprecated MongoDB
Challenge\-Response (\fBMONGODB\-CR\fP) authentication mechanism.
.sp
MongoDB adds support for SCRAM mechanism using the SHA\-256 hash
function (\fBSCRAM\-SHA\-256\fP).

.TS
center;
|l|l|.
_
T{
Value
T}	T{
Description
T}
_
T{
SCRAM\-SHA\-1
T}	T{
\fI\%RFC 5802\fP standard
Salted Challenge Response Authentication Mechanism using the SHA\-1
hash function.
T}
_
T{
SCRAM\-SHA\-256
T}	T{
\fI\%RFC 7677\fP standard
Salted Challenge Response Authentication Mechanism using the SHA\-256
hash function.
.sp
Requires featureCompatibilityVersion set to \fB4.0\fP\&.
.sp
New in version 4.0.
T}
_
T{
MONGODB\-X509
T}	T{
MongoDB TLS/SSL certificate authentication.
T}
_
T{
GSSAPI (Kerberos)
T}	T{
External authentication using Kerberos. This mechanism is
available only in \fI\%MongoDB Enterprise\fP\&.
T}
_
T{
PLAIN (LDAP SASL)
T}	T{
External authentication using LDAP. You can also use \fBPLAIN\fP
for authenticating in\-database users. \fBPLAIN\fP transmits
passwords in plain text. This mechanism is available only in
\fI\%MongoDB Enterprise\fP\&.
T}
_
.TE
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
You cannot specify both \fI\%\-\-authenticationMechanism\fP and \fI\%\-\-uri\fP\&.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiServiceName
New in version 2.6.

.sp
Specify the name of the service using GSSAPI/Kerberos\&. Only required if the service does not use the
default name of \fBmongodb\fP\&.
.sp
This option is available only in MongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiHostName
New in version 2.6.

.sp
Specify the hostname of a service using GSSAPI/Kerberos\&. \fIOnly\fP required if the hostname of a machine does
not match the hostname resolved by DNS.
.sp
This option is available only in MongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-locks
Toggles the mode of \fI\%mongotop\fP to report on use of per\-database
locks\&. This data is only available when connected to a
MongoDB 2.6 or older instance.
.sp
\fI\%\-\-locks\fP returns an error when called against a \fBmongod\fP 3.0 or
newer instance that does not report per\-database lock usage.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-rowcount int, \-n int
Number of lines of data that \fBmongotop\fP should print. "0 for indefinite"
.UNINDENT
.INDENT 0.0
.TP
.B \-\-json
New in version 3.0.0.

.sp
Returns output for \fBmongotop\fP in JSON format.
.UNINDENT
.INDENT 0.0
.TP
.B <sleeptime>
The final argument is the length of time, in seconds, that
\fBmongotop\fP waits in between calls. By default \fBmongotop\fP returns
data every second.
.UNINDENT
.SH FIELDS
.sp
When connected to a \fBmongod\fP instance, the program reports
first on the connection and then reports on the statistics at the
configured frequency. \fI\%mongotop\fP returns time values
specified in milliseconds (ms.)
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
2019\-04\-29T15:35:27.785\-0400 connected to: 127.0.0.1

             ns    total      read      write    <timestamp>
<db.collection>    81802ms     0ms    81802ms
\&...

             ns    total      read      write    <timestamp>
<db.collection>    0ms         0ms        0ms
\&...
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fI\%mongotop\fP only reports active namespaces or databases,
depending on the \fI\%\-\-locks\fP option. If you don\(aqt see a database
or collection, it has received no recent activity. You can issue a
simple operation in the \fBmongo\fP shell to generate activity to
affect the output of \fI\%mongotop\fP\&.
.INDENT 0.0
.TP
.B mongotop.ns
Contains the database namespace, which combines the database name
and collection.
.sp
If you use the \fI\%mongotop \-\-locks\fP, the \fI\%ns\fP
field does not appear in the \fI\%mongotop\fP output.
.UNINDENT
.INDENT 0.0
.TP
.B mongotop.db
Contains the name of the database. The database named \fB\&.\fP refers
to the global lock, rather than a specific database.
.sp
This field does not appear unless you have invoked
\fI\%mongotop\fP with the \fI\%\-\-locks\fP option.
.UNINDENT
.INDENT 0.0
.TP
.B mongotop.total
Provides the total amount of time that this \fBmongod\fP spent
operating on this namespace.
.UNINDENT
.INDENT 0.0
.TP
.B mongotop.read
Provides the amount of time that this \fBmongod\fP spent
performing read operations on this namespace.
.UNINDENT
.INDENT 0.0
.TP
.B mongotop.write
Provides the amount of time that this \fBmongod\fP spent
performing write operations on this namespace.
.UNINDENT
.INDENT 0.0
.TP
.B mongotop.<timestamp>
Provides a time stamp for the returned data.
.UNINDENT
.SH EXAMPLES
.sp
By default \fI\%mongotop\fP connects to the MongoDB instance
running on the localhost port \fB27017\fP\&. However, \fI\%mongotop\fP can optionally
connect to remote \fBmongod\fP
instances. See the \fI\%mongotop options\fP for more
information.
.sp
To force \fI\%mongotop\fP to return less frequently specify a number, in
seconds at the end of the command. In this example, \fI\%mongotop\fP will
return every 15 seconds.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongotop 15
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This command produces the following output:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
2019\-04\-29T15:35:27.785\-0400 connected to: 127.0.0.1


NEWS
Ticket reference tags refer to tickets in the MongoDB JIRA tracking system:
https://jira.mongodb.org

WiredTiger release 2.9.3, 2017-06-27
------------------------------------

See the upgrading documentation for details of API and behavior changes.

Significant changes:
* WT-2972       Add an interface allowing partial updates to existing values
* WT-3063       Add an interface allowing reservation of records for read-modify-write
* WT-3142       Add a workload generator application
* WT-3160       Improve eviction of internal pages from idle trees
* WT-3245       Avoid hangs on shutdown when a utility thread encounters an error
* WT-3258       Improve visibility into thread wait time due to pages exceeding memory_page_max
* WT-3263       Allow archive on restart/recovery if clean shutdown
* WT-3287       Review WiredTiger internal panic checks
* WT-3292       Review/cleanup full-barrier calls in WiredTiger
* WT-3296       LAS table fixes/improvements
* WT-3327       Checkpoints can hang if time runs backward
* WT-3345       Improve rwlock scaling
* WT-3373       Access violation due to a bug in internal page splitting
* WT-3379       Change when pages can be split to avoid excessively slowing some operations

See JIRA changelog for a full listing:
https://jira.mongodb.org/browse/WT/fixforversion/18291

WiredTiger release 2.9.2, 2017-05-25
------------------------------------

See the upgrading documentation for details of API and behavior changes.

Significant changes:
* SERVER-16796  Increase logging activity for journal recovery operations
* SERVER-28168  Cannot start or repair MongoDB after unexpected shutdown.
* SERVER-28194  Fix a bug where a missing WiredTiger.turtle file can result in data loss
* WT-98         Update the current cursor value without a search
* WT-2439       Enhance reconciliation page layout
* WT-2898       Improve performance of eviction-heavy workloads by dynamically controlling the number of eviction threads
* WT-3097       Race on reconfigure or shutdown can lead to waiting for statistics log server
* WT-3106       Add truncate support to command line wt utility
* WT-3114       Avoid archiving log files immediately after recovery
* WT-3115       Change the dhandle lock to a read/write lock
* WT-3127       Fix a bug with CPU yield calls don't necessarily imply memory barriers
* WT-3135       Fix a bug with search_near() for index with custom collator
* WT-3137       Fix a hang in __log_slot_join/__log_slot_switch_internal
* WT-3144       Fix a bug where random cursor returns not-found when descending to an empty page
* WT-3148       Improve eviction efficiency with many small trees
* WT-3149       Change eviction to start new walks from a random place in the tree
* WT-3150       Reduce impact of checkpoints on eviction server
* WT-3152       Convert table lock from a spinlock to a read write lock
* WT-3157       Fix a bug in checkpoint/transaction integrity issue when writes fail.
* WT-3158       Fix structure packing on Windows.
* WT-3188       Fix error handling in logging where fatal errors could lead to a hang
* WT-3193       Close a race between verify opening a handle and eviction visiting it
* WT-3206       Fix a bug where core dump happened on NULL page index
* WT-3218       Fix a bug that could lead to unexpected checkpoint ordering failures
* WT-3243       Reorder log slot release so joins don't wait on IO
* WT-3262       Stop making schema operations wait for cache

See JIRA changelog for a full listing:
https://jira.mongodb.org/browse/WT/fixforversion/17874


WiredTiger release 2.9.1, 2016-12-22
------------------------------------

New features and API changes; refer to the API documentation for full details:

* SERVER-26545	Remove fixed-size limitation on WiredTiger hazard pointers. See the upgrading documentation for details
* WT-283	Add a new WT_SESSION::alter method that can be used to reconfigure table metadata
* WT-2670	Change the default file system access pattern advice for data files from random to no advice. Add access_pattern_hint configuration option for WT_SESSION::create API that can be used to advise the file system of expected access semantics. See the upgrading documentation for details.
* WT-3034	Add support for including updates when reading from named snapshots

Significant changes and bug fixes:

* WT-2960	Reduce likelihood of using the lookaside file, especially when inserting multi-megabyte values
* WT-3056	Allow projected table and join cursors to use primary keys
* WT-3070	Fix a bug in search_near on indexes

Other noteworthy changes since the previous release:

* WT-2336	Add a test validating schema operations via file system call monitoring
* WT-2402	Pad structures to avoid cache line sharing
* WT-2771	Add a statistic to track per-btree dirty cache usage
* WT-2833	Add projections to wt dump utility
* WT-2969	Possible snapshot corruption during compaction
* WT-3014	Add GCC/clang support for ELF symbol visibility
* WT-3021	Fixes for java log example, raw mode in java, and raw mode in log cursors
* WT-3025	Fix error path in log_force_sync
* WT-3028	Don't check for blocked eviction with in-memory workloads
* WT-3030	Fix a race between scans and splits reading the index hint
* WT-3037	Clean up some log slot comments
* WT-3048	WiredTiger maximum size warning uses the wrong format
* WT-3051	Remove external __wt_hex symbol
* WT-3052	Improve search if index hint is wrong
* WT-3053	Make Python use internal memory allocation again
* WT-3054	Make a PackOutputStream constructor that is compatible with the previous interface.
* WT-3055	When an AsyncOp is created, cache the whether the cursor is "raw"
* WT-3057	WiredTiger hazard pointers should use the WT_REF, not the WT_PAGE
* WT-3061	Syscall testing should support pwrite64 on Linux
* WT-3064	Minor tree cleanups: .gitignore, NEWS misspelling
* WT-3066	Minor code cleanups
* WT-3068	Copy artifacts of test runs in wtperf_run script

src/mongo/gotools/mongotop/main/mongotop.go
// Main package for the mongotop tool.
package main

import (
	"github.com/mongodb/mongo-tools/common/db"
	"github.com/mongodb/mongo-tools/common/log"
	"github.com/mongodb/mongo-tools/common/options"
	"github.com/mongodb/mongo-tools/common/signals"
	"github.com/mongodb/mongo-tools/common/util"
	"github.com/mongodb/mongo-tools/mongotop"
	"gopkg.in/mgo.v2"
	"os"
	"strconv"
	"time"
)

func main() {
	// initialize command-line opts
	opts := options.New("mongotop", mongotop.Usage,
		options.EnabledOptions{Auth: true, Connection: true, Namespace: false})
	opts.UseReadOnlyHostDescription()

	// add mongotop-specific options
	outputOpts := &mongotop.Output{}
	opts.AddOptions(outputOpts)

	args, err := opts.Parse()
	if err != nil {
		log.Logvf(log.Always, "error parsing command line options: %v", err)
		log.Logvf(log.Always, "try 'mongotop --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	// print help, if specified
	if opts.PrintHelp(false) {
		return
	}

	// print version, if specified
	if opts.PrintVersion() {
		return
	}

	log.SetVerbosity(opts.Verbosity)
	signals.Handle()

	if len(args) > 1 {
		log.Logvf(log.Always, "too many positional arguments")
		log.Logvf(log.Always, "try 'mongotop --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	sleeptime := 1 // default to 1 second sleep time
	if len(args) > 0 {
		sleeptime, err = strconv.Atoi(args[0])
		if err != nil || sleeptime <= 0 {
			log.Logvf(log.Always, "invalid sleep time: %v", args[0])
			os.Exit(util.ExitBadOptions)
		}
	}
	if outputOpts.RowCount < 0 {
		log.Logvf(log.Always, "invalid value for --rowcount: %v", outputOpts.RowCount)
		os.Exit(util.ExitBadOptions)
	}

	if opts.Auth.Username != "" && opts.Auth.Source == "" && !opts.Auth.RequiresExternalDB() {
		log.Logvf(log.Always, "--authenticationDatabase is required when authenticating against a non $external database")
		os.Exit(util.ExitBadOptions)
	}

	// connect directly, unless a replica set name is explicitly specified
	_, setName := util.ParseConnectionString(opts.Host)
	opts.Direct = (setName == "")
	opts.ReplicaSetName = setName

	// create a session provider to connect to the db
	sessionProvider, err := db.NewSessionProvider(*opts)
	if err != nil {
		log.Logvf(log.Always, "error connecting to host: %v", err)
		os.Exit(util.ExitError)
	}

	if setName == "" {
		sessionProvider.SetReadPreference(mgo.PrimaryPreferred)
	}

	// fail fast if connecting to a mongos
	isMongos, err := sessionProvider.IsMongos()
	if err != nil {
		log.Logvf(log.Always, "Failed: %v", err)
		os.Exit(util.ExitError)
	}
	if isMongos {
		log.Logvf(log.Always, "cannot run mongotop against a mongos")
		os.Exit(util.ExitError)
	}

	// instantiate a mongotop instance
	top := &mongotop.MongoTop{
		Options:         opts,
		OutputOptions:   outputOpts,
		SessionProvider: sessionProvider,
		Sleeptime:       time.Duration(sleeptime) * time.Second,
	}

	// kick it off
	if err := top.Run(); err != nil {
		log.Logvf(log.Always, "Failed: %v", err)
		os.Exit(util.ExitError)
	}
}
